src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs

```diff
@@ -4739,6 +4739,12 @@ public override Task<bool> ReadAsync(CancellationToken cancellationToken)
                     return Task.FromCanceled<bool>(cancellationToken);
                 }
 
+                IDisposable registration = null;
+                if (cancellationToken.CanBeCanceled)
```

Comment:
Happy to see your contributions again @cheenamalhotra :)

I don't know enough about this code to fully understand what moving the registration does (but can look more in-depth).

However, the original reason I opened #1065 was the line just above - if IsCancellationRequested is true, ReadAsync returns immediately. That means that if the token is cancelled just before ReadAsync is invoked, it will not actually cancel the query in SQL Server, and it will keep running.

Note that if a token is in the cancelled state, [CancellationToken.Register](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.register?view=net-7.0#system-threading-cancellationtoken-register(system-action)) immediately runs the callback. That means that it may be enough to reverse the order here, and call Register *before* checking whether IsCancellationRequested is true and exiting.

Note: it's possible that your change still improves the situation and fixes the scenario in https://github.com/dotnet/SqlClient/issues/1065#issuecomment-1262671004; but as long as ReadAsync simply exits early when IsCancellationRequested is true, I don't think #1065 is fixed...

Reply:
It is nice to be around again 🙂 SqlClient is indeed my favourite team ❤️ I just got notified today and realized this issue was left out in my backlog...

I agree I missed the before call cancellation scenario here. Will move registration up further and add test-case accordingly tomorrow!

Reply:
Addressed.